### PR TITLE
gnrc_netif2: gnrc_ipv6_nib: full integration into GNRC

### PR DESCRIPTION
**Pull request by miri64**
Reintegration PR for `gnrc_netif2_integration/master` / organizational issue for GNRC integration of `gnrc_ipv6_nib` (the new neighbor discovery handler) and `gnrc_netif2` (the new network interfaces; to be renamed `gnrc_netif` before this one get's merged)

See also the [model of the NIB](https://miri64.github.io/riot-ndp-model/) for more information.

**!!! THIS PR DOES NOT REQUIRE CODE REVIEW, ONLY TESTING !!! IT IS AN AGGREGATION OF THE PULL REQUESTS MENTIONED BELOW!!!**

# TODOs

(in order of priority)

- [x] various fixes:
    - [x] #7910 
    - [x] #7911 
    - [x] #7926
    - [x] #7935
    - [x] #7960
    - [x] #7961
    - [x] #7966
    - [x] #8016 
    - [x] #8041 
    - [x] #8063
    - [x] #8071
    - [x] #8076
    - [x] #8082
    - [x] #8087
- [x] merge `gnrc_netif2` integration (@miri64, #7424) [until Nov 3]
- [x] merge `gnrc_ipv6_nib` port to `gnrc_netif2` (@miri64, #7456) [until Nov 3]
- [x] merge router discovery part of `gnrc_ipv6_nib` (@miri64, #7479) [until Nov 15]
- [x] adapt NIB for route timeouts (@cgundogan + @miri64, #8073)
- [x] port RPL to `gnrc_ipv6_nib` (@cgundogan, #8010) [until Nov 20]
- [x] port LWMAC to `gnrc_netif2` (@zhuoshuguo, #7895)
- [x] merge `xbee` adaption layer for `gnrc_netif2` (@miri64, #7898) [until Nov 15]
- [x] merge raw adaption layer for `gnrc_netif2` (@miri64, #7462) [until Nov 15]
- [x] merge `cc110x` adaption layer for `gnrc_netif2` (@miri64, #7414) [until Nov 15]
- [x] merge Nordic softdevice adaption layer for `gnrc_netif2` (@miri64, #7419) [until Nov 15]
- [x] Adapt 6ctx shell command for `gnrc_ipv6_nib` (@miri64, #8077) [until Nov 20]
- [x] Remove all old `gnrc_netif` and `gnrc_ndp` (@miri64, #8064) [until Nov 20]
- [x] `s/2$//g` (@miri64, #8069) [until Nov 20]

This PR is part of the network layer remodelling effort:
PR dependencies

### CODE REVIEW

**Comment by miri64** [comment]
Ping @RIOT-OS/maintainers? We are already three days behind schedule.

**Comment by aabadie** [comment]
@miri64, I want to adapt #7356 to work with netif2. Should I rebase my initial branch on `gnrc_netif2_integration/master` and open another PR on it ?

**Comment by miri64** [comment]
@kaspar030 There is some [unrelated bug](https://ci.riot-os.org/RIOT-OS/RIOT/7925/b1e69a776fbdc546557c242509654d6292ea1cdf/output.html) with JavaScript in this PR. I don't want to rebase this PR to master. Any idea what is happening?

**Comment by miri64** [comment]
Mhhh.. to my information `sys/net/gnrc/application_layer/uhcpc/gnrc_uhcpc.c` was changed [last on Oct 12 to my knowledge](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/application_layer/uhcpc/gnrc_uhcpc.c) so before the embargo... so why is there a conflict now oO?

**Comment by miri64** [comment]
Maybe this issue resolves itself later down the line, otherwise we need to find a solution...

**Comment by miri64** [comment]
> Mhhh.. to my information sys/net/gnrc/application_layer/uhcpc/gnrc_uhcpc.c was changed last on Oct 12 to my knowledge so before the embargo... so why is there a conflict now oO?

That was true, but #7722 which introduced that commit was merged during the embargo => it is what caused the conflict. There is a backport in #7960 that should resolve it.

**Comment by bergzand** [comment]
I noticed that the tests for the enc28j60 and the encx24j600 still fail (with #7456). To me it looks like a test that isn't ported to netif2. Is there already a PR for fixing those tests? (Not sure if I'm just missed something obvious here)

**Comment by miri64** [comment]
I'll look into that and fix them in #7966 if this is the case.

**Comment by miri64** [comment]
@bergzand fix provided in https://github.com/RIOT-OS/RIOT/pull/7966/commits/a3664c923878a8d573122eac82e3e25d515176e9

**Comment by bergzand** [comment]
And there is the next merge conflict -_-

**Comment by miri64** [comment]
> And there is the next merge conflict -_-

Fixed in #8009.

**Comment by miri64** [comment]
\>50% of merging done :dancer:

**Comment by miri64** [comment]
@smlng @PeterKietzmann @cgundogan @bergzand @biboc @haukepetersen @A-Paul and anyone who is also interested in testing the shit out of this: Now that #7479 is merged there are no further plans to add more complex functionality. @cgundogan is currently working on RPL (which will involve a slight change to the FT view of the NIB) and I expect it to be in this branch tomorrow or monday at the latest. My next steps will be to remove the legacy code (see #8064) and rename all `gnrc\(*\)2` stuff to `gnrc\1` (ongoing). So it's time to think of some crazy IoT scenarios and run them with this branch ;-).

**Comment by miri64** [comment]
Also, if you have time have a look if you can review some of the remaining PRs. I usually note them down in the list above and add them to [the project](https://github.com/RIOT-OS/RIOT/projects/6) (the tree is a little bit time-wastey to maintain, but the PRs in it are the oldest, so they have somewhat of a priority).

**Comment by miri64** [comment]
Gnarrrrfff there was a conflict again. Will investigate in a moment.

**Comment by miri64** [comment]
Revert to `tests/gnrc_netif2` in master in #8072.

**Comment by miri64** [comment]
:( #8072 did not solve the conflict... What do we do now? Merge #7979 into master?

**Comment by smlng** [comment]
>:( #8072 did not solve the conflict... What do we do now? Merge #7979 into master this PR?

that's a good example why I feel the GNRC embargo is somewhat _problematic_ ... 

In the end we start (and already did so) to revert changes and reintroduce them on master just to keep away conflicts from this branch. But what we really get (or got) out of it, is a messy git log/commit history of our master branch ... 

IMHO: rebase this branch onto latest master, resolve conflicts and repeat as needed until feature is completed here, and is ready to be finally merged into master.

**Comment by aabadie** [comment]
\<off topic comment\>
> But what we really get (or got) out of it, is a messy git log/commit history of our master branch ...

Indeed, and probably because of the merge commits. GitHub allows to choose between 3 kinds of actions:
* merge => create a merge commit
* rebase and merge => rebase the full history of commits of a PR on top of latest master
* squash and merge => squash all commits of a PR into a single one and rebase it on top of master

Using solution 2 in our workflow would certainly fix the messy git history.

\</off topic comment\>

**Comment by smlng** [comment]
>  and probably because of the merge commits

@aabadie No that's not what I meant, the merge commits can certainly be omitted by using the `rebase+merge` option in Github, but then we loose info who merged which PR.

What I was referring to here was, that recently some changes on master were reverted, because they introduced conflicts with this branch (see #8009, #7967, and others). This clutters the git history with 2 commits more than actually needed, namely: org-commit, revert, re-commit (after feature merge).

**Comment by miri64** [comment]
> that's a good example why I feel the GNRC embargo is somewhat problematic ...

But I still did not here any constructive solution that is better from anyone...

> In the end we start (and already did so) to revert changes and reintroduce them on master just to keep away conflicts from this branch.

Well the ideal situation would have been if people would have paid more attention... But things happened now.


> But what we really get (or got) out of it, is a messy git log/commit history of our master branch ...

Our git log is messy anyways, because we *always* merge PRs and do not rebase them....

> IMHO: rebase this branch onto latest master, resolve conflicts and repeat as needed until feature is completed here, and is ready to be finally merged into master.

So you want to review the code then?

Anyway, I think the best solution now is to merge #7979 into this branch as well.

**Comment by miri64** [comment]
For the record: Wasn't the sanest idea after all -.-... will revert this

**Comment by miri64** [comment]
I reverted https://github.com/RIOT-OS/RIOT/pull/8074 by directly pushing into this branch, just to not further poison the git history. abe721f is still the same commit that was created in #8071, so this is still in line with my "don't change history" approach to this PR.

... But we still have no way of resolving the merge conflict yet -.-

**Comment by miri64** [comment]
\o/ fix is ugly (and doesn't help fight having a weird git history), but problem solved.

**Comment by miri64** [comment]
> What I was referring to here was, that recently some changes on master were reverted, because they introduced conflicts with this branch (see #8009, #7967, and others). This clutters the git history with 2 commits more than actually needed, namely: org-commit, revert, re-commit (after feature merge).

@smlng BTW. That argument basically applies also to any PR against master that was wrongly merged and needed to be reverted (and than maybe re-merged in a fixed state afterwards). That is and always has been our approach to deal with improper review: If you do not review according to the rules, your change needs to be reverted. Since the "letter of the law" at the moment currently includes "don't merge anything remotely related to GNRC" this applies also for #7979 and #8009.

**Comment by miri64** [comment]
Now only the RPL port is missing :-).

**Comment by miri64** [comment]
With RPL now merged, this PR isn't WIP anymore. There are some bugs already reported that needs to be fixing (will do that tomorrow ASAP), but apart from that: This is ready to test. Due to the size of this PR, I think it's a good idea to have 3 tested (!) ACKs for this PR, before merging. However, we also need to merge this ASAP to lift the GNRC embargo, so if we don't get all three together until the end of the week I can live with less.

**Comment by zhuoshuguo** [comment]
Just tested `examples/gnrc_networking`, it seems that udp transmission doesn't work.

**Comment by miri64** [comment]
> Just tested `examples/gnrc_networking`, it seems that udp transmission doesn't work.

Which board? From which kind of node to which?

**Comment by zhuoshuguo** [comment]
Tested on samr21-xpro, between two samr21-xpro nodes (both started a udp server).

**Comment by miri64** [comment]
I basically did the same today, just with ICMPv6 and it worked. Since this PR doesn't touch UDP, I find it highly unlikely, that it won't. I'll test tomorrow, but please re-check your set-up, including that you are on the most current version of this branch.

**Comment by bergzand** [comment]
I have 2 hardware nodes (one LR, one LBR) running without problems for 14 hours. So far everything seems to work stable and no triggered assertions. I'm not using RPL so far, I want to start testing that next for a few hours. It could be that the assertion reported in #8099 is related to RPL as I had RPL initialized and running at that moment (I think).

**TL;DR**: Networking is still stable and working after half a day! 😄

**Comment by miri64** [comment]
> TL;DR: Networking is still stable and working after half a day! :smile:

Yay, but the really interesting thing would also be to have a multihop setup tested (with and without RPL). Since you already said, that you will run RPL, I can do the non-RPL one and report my findings / setup so you can just reproduce.

**Comment by bergzand** [comment]
@miri64 I have a setup with 3 nodes running RPL for the routing. With some help of the l2filter, I have a chain of 6LBR-6LR-6LR. I fear I'm starting to run into a number of RPL related bugs (expect more bug reports (and maybe even fixes) for gnrc_rpl tomorrow).

**Comment by miri64** [comment]
> @miri64 I have a setup with 3 nodes running RPL for the routing. With some help of the l2filter, I have a chain of 6LBR-6LR-6LR. I fear I'm starting to run into a number of RPL related bugs (expect more bug reports (and maybe even fixes) for gnrc_rpl tomorrow).

For this PR to get merged it is important to verify if they exist due to the integration of the NIB+Netif(2) or exist also in master. However, bug reports are always welcome of course :-).

**Comment by miri64** [comment]
My setup (6LBR[gnrc_border_router]-6LR[gnrc_networking]-6LN[gcoap] with one static route to 6LN via 6LR at the 6LBR and a CoAP GET[CON] to 6LN every .5 sec from host system) ran apart from a few hick-ups in the bootstrapping phase for 16h straight. The hickups were:

1. GUA is not always registered to upstream router's NC immediately but takes sometimes a second try
2. 6LN isn't immediately reachable from host system but takes a few seconds

(1) might be an issue, but I see it as low priority, since after a few seconds the network repairs itself, (2) could hint towards a problem with the auto-config ethos+uhcpd is doing and thus not related to this PR.

**Comment by miri64** [comment]
@bergzand I take it as a good sign, that I did not hear from you yet :D

**Comment by bergzand** [comment]
@miri64 Same setup as [before](https://github.com/RIOT-OS/RIOT/pull/7925#issuecomment-346504695). No real network threatening issues found, everything still worked fine in the morning and almost no lost packets during the night. No full packet buffer errors and no triggered assertions :)

Only noticed issue #8133 on my nodes, but that didn't cause any real problems.

**Comment by miri64** [comment]
Now that we tested everything pretty extensively and all bugs are fixed, I think this is ready to be merged ;-)

**Comment by bergzand** [comment]
I'm not going to assume that we covered every corner case with the tests, but I'm confident enough to say that for the majority of the cases this should just work(tm).

This has my ACK, but I would really like this to have an additional ACK from somebody who is reviewing and testing code for a bit longer than I am.

**Comment by miri64** [comment]
So is anyone willing to ACK this now? This would lift the GNRC embargo. ;-)

**Comment by smlng** [comment]
> #7713 is the only PR blocking this PR. I

that's my tracking issue, I assume you meant #8147? Which I just merged, lets wait for murdock to finish once again.

**Comment by miri64** [comment]
Yes. That was wrong copy-pasta ^^"

**Comment by bergzand** [comment]
Al green!

**Comment by miri64** [comment]
I'm not hitting that button. The honor belongs to one of you guys ;-).

**Comment by bergzand** [comment]
There you go :)

🎉
